lib/Support/JSON.cpp

```diff
@@ -0,0 +1,134 @@
+//===- Json.cpp - Json Utilities --------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Utilities for file system path handling, supplementing the ones from
+// llvm::sys::path.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/JSON.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace json = llvm::json;
+
+using namespace circt;
+using mlir::UnitAttr;
+
+// NOLINTBEGIN(misc-no-recursion)
+LogicalResult circt::convertAttributeToJSON(llvm::json::OStream &json,
+                                            Attribute attr) {
+  return TypeSwitch<Attribute, LogicalResult>(attr)
+      .Case<DictionaryAttr>([&](auto attr) {
+        json.objectBegin();
+        for (auto subAttr : attr) {
+          json.attributeBegin(subAttr.getName());
+          if (failed(convertAttributeToJSON(json, subAttr.getValue())))
+            return failure();
```

Comment:
The failure path leaves the JSON object open, causing an assertion failure.  Found while poking at the unit test to trigger failures:

```
Note: This is test shard 1 of 21.
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from JSONTest
[ RUN      ] JSONTest.RoundTripTypes
/home/will/src/sifive/circt/unittests/Support/JSONTest.cpp:71: Failure
Value of: succeeded(convertAttributeToJSON(json, sampleDict))
  Actual: false
Expected: true
CIRCTSupportTests: /home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/JSON.h:965: llvm::json::OStream::~OStream(): Assertion `Stack.size() == 1 && "Unmatched begin()/end()"' failed.
 #0 0x00000000003feeab llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/will/src/sifive/circt/llvm/llvm/lib/Support/Unix/Signals.inc:602:13
 #1 0x00000000003fd112 llvm::sys::RunSignalHandlers() /home/will/src/sifive/circt/llvm/llvm/lib/Support/Signals.cpp:105:18
 #2 0x00000000003ff55d SignalHandler(int) /home/will/src/sifive/circt/llvm/llvm/lib/Support/Unix/Signals.inc:0:3
 #3 0x00007faf8da3dbf0 __restore_rt (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x3dbf0)
 #4 0x00007faf8da8abc7 __pthread_kill_implementation (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x8abc7)
 #5 0x00007faf8da3db46 gsignal (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x3db46)
 #6 0x00007faf8da284b5 abort (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x284b5)
 #7 0x00007faf8da283d9 _nl_load_domain.cold.0 (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x283d9)
 #8 0x00007faf8da367b6 (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x367b6)
 #9 0x00000000003471ca llvm::json::OStream::~OStream() /home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/JSON.h:966:5
#10 0x00000000003471ca (anonymous namespace)::JSONTest_RoundTripTypes_Test::TestBody() /home/will/src/sifive/circt/unittests/Support/JSONTest.cpp:90:1
#11 0x000000000040e6dc testing::internal::UnitTestImpl::os_stack_trace_getter() /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/src/gtest.cc:5635:7
#12 0x000000000040e6dc testing::Test::Run() /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/src/gtest.cc:2515:9
#13 0x000000000040fe19 testing::TestInfo::Run() /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/src/gtest.cc:0:0
#14 0x0000000000410751 testing::TestSuite::Run() /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/src/gtest.cc:2815:44
#15 0x000000000041e327 testing::internal::UnitTestImpl::RunAllTests() /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/src/gtest.cc:5337:24
#16 0x000000000041dc16 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/src/gtest.cc:0:0
#17 0x000000000041dc16 testing::UnitTest::Run() /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/src/gtest.cc:4925:10
#18 0x0000000000400d1a RUN_ALL_TESTS() /home/will/src/sifive/circt/llvm/third-party/unittest/googletest/include/gtest/gtest.h:0:46
#19 0x0000000000400d1a main /home/will/src/sifive/circt/llvm/third-party/unittest/UnitTestMain/TestMain.cpp:55:10
#20 0x00007faf8da2924e __libc_start_call_main (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x2924e)
#21 0x00007faf8da29309 __libc_start_main@GLIBC_2.2.5 (/nix/store/yzjgl0h6a3qh1mby405428f16xww37h0-glibc-2.35-224/lib/libc.so.6+0x29309)
#22 0x0000000000346155 _start (/home/will/src/sifive/circt/build/unittests/Support/./CIRCTSupportTests+0x346155)
```

Please handle + add a failure test case for at least some of these.  The JSON class says missing begin/end pairs is an assertion failure but when asserts are off is explicitly undefined behavior.

(Either close on the failure paths or return failure after closing the object/attribute/etc.?)

FWIW there are variants that take lambdas, for your consideration (sometimes can help re:ensuring every begin has an end, although that doesn't help with this AFAICT).